src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -5572,3 +5581,65 @@ TaskListHighestTaskId(List *taskList)
 
 	return highestTaskId;
 }
+
+
+/*
+ * ModifyProblematicNodes takes an input rewritten query and modifies
+ * nodes which, after going through our planner, pose a problem when
+ * deparsing. So far we have two such type of Nodes that may pose problems:
+ * RelabelType and CoerceIO nodes.
+ * Details will be written in comments in the corresponding if conditions.
+ */
+static Node *
+ModifyProblematicNodes(Node *inputNode)
+{
+	if (inputNode == NULL)
+	{
+		return NULL;
+	}
+
+	if (IsA(inputNode, RelabelType) &&
+		OidIsValid(((RelabelType *) inputNode)->resultcollid) &&
+		((RelabelType *) inputNode)->resultcollid != DEFAULT_COLLATION_OID)
+	{
+		/*
+		 * The planner converts CollateExpr to RelabelType
+		 * and here we convert back.
+		 */
+		return (Node *) RelabelTypeToCollateExpr((RelabelType *) inputNode);
+	}
+	else if (IsA(inputNode, CoerceViaIO))
+	{
+		/*
+		 * The planner converts some ::text/::varchar casts to ::cstring
+		 * and here we convert back to text because cstring is a pseudotype
+		 * and it cannot be casted to most resulttypes
+		 */
+
+		CoerceViaIO *iocoerce = (CoerceViaIO *) inputNode;
+		Node *arg = (Node *) iocoerce->arg;
+
+		if (IsA(arg, Const) && ((Const *) arg)->consttype == CSTRINGOID)
+		{
+			Const *cstringToText = (Const *) arg;
+
+			cstringToText->consttype = TEXTOID;
+			cstringToText->constlen = -1;
+
+			Type textType = typeidType(TEXTOID);
```

Comment:
what about NULL values? PG seems to do something different:

```C
		/*
		 * We assume here that UNKNOWN's internal representation is the same
		 * as CSTRING.
		 */
		if (!con->constisnull)
			newcon->constvalue = stringTypeDatum(baseType,
												 DatumGetCString(con->constvalue),
												 inputTypeMod);
		else
			newcon->constvalue = stringTypeDatum(baseType,
												 NULL,
												 inputTypeMod);
```

I think some of our tests should trigger NULL values (also good to check for relable types for NULL values? Would we crash?

Reply:
Or, maybe should we use makeConst instead of stringTypeDatum?

Reply:
> Or, maybe should we use makeConst instead of stringTypeDatum?

`makeConst` already assumes we are passing a correct Datum for the text type, so we need to use `stringTypeDatum` to obtain a correct Datum for text type.

Reply:
> also good to check for relable types for NULL values

If arg is NULL it will be handled from the CollateExpr node since we just pass that arg. 
`collateExpr->arg = relabelType->arg;`
This if statement already keeps us safe from any other NULL/invalidOid cases:
```C
if (IsA(inputNode, RelabelType) &&
    OidIsValid(((RelabelType *) inputNode)->resultcollid) &&
    ((RelabelType *) inputNode)->resultcollid != DEFAULT_COLLATION_OID)
```